core/lib/dal/src/transactions_dal.rs

```diff
@@ -968,34 +968,64 @@ impl TransactionsDal<'_, '_> {
-    /// during the previous execution.
-    pub async fn get_miniblocks_to_reexecute(&mut self) -> Vec<MiniblockReexecuteData> {
-        let transactions_by_miniblock: Vec<(MiniblockNumber, Vec<Transaction>)> = sqlx::query_as!(
-            StorageTransaction,
-            "SELECT * FROM transactions \
-            WHERE miniblock_number IS NOT NULL AND l1_batch_number IS NULL \
-            ORDER BY miniblock_number, index_in_block",
-        )
-        .fetch_all(self.storage.conn())
-        .await
-        .unwrap()
-        .into_iter()
-        .group_by(|tx| tx.miniblock_number.unwrap())
-        .into_iter()
-        .map(|(miniblock_number, txs)| {
-            (
-                MiniblockNumber(miniblock_number as u32),
-                txs.map(Transaction::from).collect::<Vec<_>>(),
-            )
-        })
-        .collect();
+    async fn get_miniblock_with_transactions_for(
+        &mut self,
+        mode: MiniblockExecutionMode,
+    ) -> Vec<(MiniblockNumber, Vec<Transaction>)> {
+        let transactions = match mode {
+            MiniblockExecutionMode::Reexecute => {
+                sqlx::query_as!(
+                    StorageTransaction,
+                    "SELECT * FROM transactions \
+                    WHERE miniblock_number IS NOT NULL AND l1_batch_number IS NULL \
+                    ORDER BY miniblock_number, index_in_block",
+                )
+                .fetch_all(self.storage.conn())
+                .await
+            }
+            MiniblockExecutionMode::L1Batch(l1_batch_number) => {
+                sqlx::query_as!(
+                    StorageTransaction,
+                    "SELECT * FROM transactions \
+                    WHERE l1_batch_number = $1 \
+                    ORDER BY miniblock_number, index_in_block",
+                    l1_batch_number.0 as i64,
+                )
+                .fetch_all(self.storage.conn())
+                .await
+            }
+        };
+        transactions
+            .unwrap()
+            .into_iter()
+            .group_by(|tx| tx.miniblock_number.unwrap())
+            .into_iter()
+            .map(|(miniblock_number, txs)| {
+                (
+                    MiniblockNumber(miniblock_number as u32),
+                    txs.map(Transaction::from).collect::<Vec<_>>(),
+                )
+            })
+            .collect()
+    }
+
+    /// Returns miniblocks with their transactions to be used in VM execution.
+    /// `miniblock_execution_mode` parameter describes what miniblocks are collected:
+    ///    - Reexecution -- miniblocks state_keeper needs to reexecute on restart.
+    ///      These are the transactions that are included to some miniblock,
+    ///      but not included to L1 batch. The order of the transactions is the same as it was
+    ///      during the previous execution.
+    ///    - L1Batch -- miniblocks that are included in the given L1 batch.
+    pub async fn get_miniblocks_to_execute_for(
```

Comment:
Nit: This method could propagate DB errors as well, although it could require some effort to fix its uses.